src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs

```diff
@@ -253,7 +264,86 @@ internal static unsafe byte[] UnsafeDecode(ReadOnlySpan<char> strSpan)
+            // incoming chars directly to DecodeFromUtf8InPlace. Otherwise, rent an array, copy all the
+            // data into it, and do whatever fixups are necessary on that copy, then pass that copy into
+            // DecodeFromUtf8InPlace.
+
+            const int StackAllocThreshold = 512;
+            char[] arrayPoolChars = null;
+            scoped Span<char> charsSpan = default;
+            scoped ReadOnlySpan<char> source = strSpan;
+
+            if (needReplace || decodedLength != source.Length)
+            {
+                charsSpan = decodedLength <= StackAllocThreshold ?
+                    stackalloc char[StackAllocThreshold] :
+                    arrayPoolChars = ArrayPool<char>.Shared.Rent(decodedLength);
+                charsSpan = charsSpan.Slice(0, decodedLength);
+
+                source = HandlePaddingAndReplace(needReplace, charsSpan, source);
+            }
+
+            byte[] arrayPoolBytes = null;
+            Span<byte> bytesSpan = decodedLength <= StackAllocThreshold ?
+                stackalloc byte[StackAllocThreshold] :
+                arrayPoolBytes = ArrayPool<byte>.Shared.Rent(decodedLength);
+
+            int length = Encoding.UTF8.GetBytes(source, bytesSpan);
+            Span<byte> utf8Span = bytesSpan.Slice(0, length);
+            OperationStatus status = System.Buffers.Text.Base64.DecodeFromUtf8InPlace(utf8Span, out int bytesWritten);
+            Debug.Assert(status == OperationStatus.Done, "Expected DecodeFromUtf8 to be successful");
+
+            utf8Span.Slice(0, bytesWritten).CopyTo(output);
+
+            if (arrayPoolBytes is not null)
+            {
+                bytesSpan.Clear();
+                ArrayPool<byte>.Shared.Return(arrayPoolBytes);
+            }
+
+            if (arrayPoolChars is not null)
+            {
+                charsSpan.Clear();
+                ArrayPool<char>.Shared.Return(arrayPoolChars);
+            }
+        }
+
+        private static unsafe ReadOnlySpan<char> HandlePaddingAndReplace(bool needReplace, Span<char> charsSpan, ReadOnlySpan<char> source)
+        {
+            source.CopyTo(charsSpan);
+            if (source.Length < charsSpan.Length)
+            {
+                charsSpan[source.Length] = base64PadCharacter;
+                if (source.Length + 1 < charsSpan.Length)
+                {
+                    charsSpan[source.Length + 1] = base64PadCharacter;
+                }
+            }
+
+            if (needReplace)
+            {
+                Span<char> remaining = charsSpan;
+                int pos;
+                while ((pos = remaining.IndexOfAny(base64UrlCharacter62, base64UrlCharacter63)) >= 0)
+                {
+                    remaining[pos] = (remaining[pos] == base64UrlCharacter62) ? base64Character62 : base64Character63;
+                    remaining = remaining.Slice(pos + 1);
+                }
```

Comment:
This does not seem optimal. Wouldnt it end up being a O(N^2) in the edge case? Simple for loop with replacing in place would be more optimal.

E.g.

for (int i = 0; i < remaining.Length; i++)
{
    if (remaining[i] == base64UrlCharacter62)
    {
        remaining[i] = base64Character62;
    }
    else if (remaining[i] == base64UrlCharacter63)
    {
        remaining[i] = base64Character63;
    }
}

Reply:
I don't follow, how would it be n^2? We search, we replace, we continue searching from our last position.